### ISSUE TITLE
Add support to NestedDiff for calculating differences between trees of Elements of arbitrary depth?

### ISSUE THREAD

**Issue by tonyarnold** [wokalski/Diff.swift#34]
This might be wishful thinking, but I'm currently looking at support for a data model structure that backs an NSOutlineView. At the moment, this is a tree of elements, like so:

```
Element
    -> Element
    -> Element
    -> Element
    -> Element
         -> Element
```

And so on. Basically Elements can contain 0 or more Elements, and there's no real limit on the depth of the tree.

I'm no algorithmic whiz kid, but this seems like it's not outside the realm of possibility using this library. I realise there are performance implications to traversing and calculating differences in a structure like this. Putting that aside for a moment, is it worth pursuing?

**Comment by wokalski** [comment]
I think it is worth pursuing, sure. The first problem from the top of my had is type system. I am not swift whiz kid but it seems harrd, is it even possible?

**Comment by tonyarnold** [comment]
Not sure. I think there would need to be a `TreeDiffable` protocol that Elements would need to conform to, and it could provide some basic info similar to `NSTreeNode` (https://developer.apple.com/reference/appkit/nstreenode#):

```swift
protocol TreeDiffable: Equatable {
  var hasChildren: Bool { get }
  var children: [TreeDiffable] { get }
  var parent: TreeDiffable? { get }
}
```

The hard part would be constructing the IndexPaths for objects that move between parents. Everything else should be pretty similar to what you're doing now. I think.

**Comment by wokalski** [comment]
Aha, right.

So I think it's doable. Recursively go through the tree and pass the current path in an array. It won't be efficient but good first step. I'm not sure if it's needed to have parent and/or indexInParent.

**Comment by tonyarnold** [comment]
Yeah, the index in parent getter probably isn't necessary given you can look that up as you need it. 

The main reason I included the parent as a property of the child is that `NSOutlineView` expects you to pass that with most actions you perform on an item (it works on items and their parents, not indexes like NSTableView).

**Comment by wokalski** [comment]
Bingo. So we can pass the parent during recursion.

**Comment by tonyarnold** [comment]
Unless you're super keen to do it, I'll give this some thought and have a go at a rough first cut for this change. Thanks for the advice!

**Comment by wokalski** [comment]
awesome, do it!
